src/coreclr/pal/src/CMakeLists.txt

```diff
@@ -308,6 +308,10 @@ if(CLR_CMAKE_TARGET_LINUX)
     target_link_libraries(coreclrpal PUBLIC ucontext)
   endif(CLR_CMAKE_TARGET_LINUX_MUSL AND (CLR_CMAKE_TARGET_ARCH_I386 OR CLR_CMAKE_TARGET_ARCH_POWERPC64))
 
+  if(CLR_CMAKE_TARGET_ARCH_I386)
+      target_link_libraries(coreclrpal PUBLIC atomic)
```

Comment:
Why do we need an extra library now?

It contradicts the "These functions are implemented in terms of the ‘__atomic’ builtins" statement.

Reply:
FWIW @MichalPetryka did some tests in the past with Godbolt and various compilers. The non-dependency on the atomic library was one of the reasons for using the `__sync*` APIs. Obviously, the situation may have changed since then, or errors could have been done in the initial assessment, but the addition of the external library is a red flag.

Reply:
It is only needed for x86, because toolchain fallback to library for x86.

Reply:
It does not fallback to library with `__sync*` APIs. It means that the implementation of at least some of the `__atomic` builtins is orthogonal to `__sync*` APIs.

Reply:
Yup, arm32 has atomic instructions like `LDREX` and `STREX` etc. for 64-bit operations, so compiler built-in __sync* and C11 __atomic directly map to those. i386 case is so that toolchain calls library for 64-bit __atomic ops and directly emits assembly instructions for __sync ops, in order to keep their __atomic implementation simple (while conforming to std=c11).

Also, I think, the proper way is to try-until-success for both __sync and __atomic APIs, like we are doing in: https://github.com/dotnet/runtime/blob/13737d2d4141d78faa0712ce8ec7d018523ca3f5/src/coreclr/pal/inc/pal.h#L3547-L3549